selfhost/typechecker.jakt

```diff
@@ -3759,20 +3780,65 @@ struct Typechecker {
         if .inside_defer {
             .error("‘return’ is not allowed inside ‘defer’", span)
         }
+
+        mut type_hint: TypeId? = None
+        if .current_function_id.has_value() {
+            type_hint = Some(.get_function(.current_function_id!).return_type_id)
+        }
+
         if not expr.has_value() {
+            // Return expression is empty, so we try to return void
+
+            if type_hint.has_value() and not type_hint!.equals(void_type_id()){
+                // We didn't expect to return void
+                .error(
+                    format("Type mismatch: expected to return `{}`, but got `{}`",
+                    .type_name(type_hint!),
+                    .type_name(void_type_id())
+                    ),
+                span)
+            }
+
+            if (.current_function_id.has_value() and .get_function(.current_function_id!).name == "main") {
+                // We can't return void in the `main` function
+                .error("Main function must return c_int", span)
+            }
+
             return CheckedStatement::Return(val: None, span)
         }
 
         if not (.current_function_id.has_value() and .get_function(.current_function_id!).is_comptime) and expr! is Function {
             .error("Returning a function is not currently supported", span)
         }
 
-        mut type_hint: TypeId? = None
-        if .current_function_id.has_value() {
-            type_hint = Some(.get_function(.current_function_id!).return_type_id)
+        let checked_expr = .typecheck_expression(expr!, scope_id, safety_mode, type_hint)
+
+        if .current_function_id.has_value() and .get_function(.current_function_id!).name == "main" {
+            // Even without a type_hint, "main" function has a c_int return type, we can only return an integer
+            if not .is_integer(checked_expr.type()) {
+                .error("Main function must return c_int", checked_expr.span())
+            }
+        }
+
+        if not type_hint.has_value()
+        or type_hint.value().equals(unknown_type_id())
+        or .get_type(checked_expr.type()) is TypeVariable
+        // FIXME: Not sure how, but we should deal with GenericInstance differently.
+        // Ex. in tests/typechecker/generic_across_scopes.jakt we don't check the return type of make_memo properly
+        or .get_type(type_hint!) is GenericInstance {
+            // We don't yet exactly know the type of what we need to return / type of what are we returning
+            return CheckedStatement::Return(val: checked_expr, span)
+        }
+
+        if not type_hint.value().equals(checked_expr.type()) {
```

Comment:
Don't do this, please never use `TypeId::equals` to check for type compatibility - there's a function for that, it even logs the errors for you.

Reply:
I can see why :P
After I changed it to `check_types_for_compat`, the early return in case line 3828 `or .get_type(type_hint!) is GenericInstance` is true, is no longer needed. Unfortunately it doesn't solve the problem of make_memo so the FIXME belongs somewhere else (maybe TypeVariable?).

This allows us to properly check the return type of the function "find" in `samples/closures/generic_closure_with_parameter.jakt`

```cpp
function find<T>(anon arr: [T], anon cb: function(a: T) -> bool) -> T? {
    for i in 0..arr.size() {
        if cb(a: arr[i]) {
            return i
        }
    }
    return None
}

function main() {
    mut arr = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
    
    println("{}", find(arr, function(a: i64) -> bool => a == 5).value_or(-1))
}
```

`return i` gives an error now, because we try to return `usize` instead of `i64?`. I'll change it to `return i as! i64` and add a test case for the error.